File: flutter/test_app/lib/data/supabase/network/service/supabase_auth_service.dart
import 'package:supabase_flutter/supabase_flutter.dart';

import '../../../../domain/model/user_role.dart';


class SupabaseAuthService {
  final SupabaseClient client;
  SupabaseAuthService(this.client);

  Future<void> signIn({
    required String email,
    required String password,
  }) async {
    final res = await client.auth.signInWithPassword(
      email: email,
      password: password,
    );

    if (res.user == null) {
      throw AuthException('Login failed: user is null');
    }
  }

  Future<void> signOut() async {
    await client.auth.signOut();
  }

  String? currentUserId() => client.auth.currentUser?.id;

  Future<UserRole> getCurrentUserRole() async {
    final userId = currentUserId();
    if (userId == null) return UserRole.SPORTAS;
    final trenerRow = await client
        .from('Trener')
        .select('IdKorisnika')
        .eq('IdKorisnika', userId)
        .maybeSingle();

    if (trenerRow != null) return UserRole.TRENER;

    final sportasRow = await client
        .from('Sportas')
        .select('IdKorisnika')
        .eq('IdKorisnika', userId)
        .maybeSingle();

    if (sportasRow != null) return UserRole.SPORTAS;

    return UserRole.SPORTAS;  }
}

File: kotlin_jetpack_compose/app/src/main/java/com/example/testapp/data/supabase/network/service/SupabaseAuthService.kt
package com.example.testapp.data.supabase.network.service


import com.example.testapp.domain.model.UserRole
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.auth.providers.builtin.Email
import io.github.jan.supabase.postgrest.from
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.serialization.Serializable


class SupabaseAuthService(
    private val client: SupabaseClient
) {
    @Serializable private data class SportasRow(val IdKorisnika: String)
    @Serializable private data class TrenerRow(val IdKorisnika: String)

    suspend fun signIn(email: String, password: String) {
        client.auth.signInWith(Email) {
            this.email = email
            this.password = password
        }
    }

    suspend fun signOut() {
        client.auth.signOut()
    }

    fun currentUserId(): String? = client.auth.currentUserOrNull()?.id

    suspend fun getCurrentUserRole(): UserRole {
        val userId = currentUserId() ?: return UserRole.SPORTAS

        val isTrener = client.from("Trener")
            .select { filter { eq("IdKorisnika", userId) } }
            .decodeList<TrenerRow>()
            .isNotEmpty()

        if (isTrener) return UserRole.TRENER

        val isSportas = client.from("Sportas")
            .select { filter { eq("IdKorisnika", userId) } }
            .decodeList<SportasRow>()
            .isNotEmpty()

        if (isSportas) return UserRole.SPORTAS

        return UserRole.SPORTAS
    }
}
